modules/dcache-webdav/src/main/java/org/dcache/webdav/transfer/CredentialServiceClient.java

```diff
@@ -164,48 +156,58 @@ public StaticOpenIdCredential getDelegatedCredential(String token,
                             secret));
 
                 return createOidcCredential(host, id, secret, json);
-            } catch (AuthenticationException | IOException | JSONException e) {
-                failures.add(String.format("[%s -> %s]", host, e.toString()));
+            } catch (IOException | JSONException e) {
+                failures.add(String.format("[%s -> %s]", host, e));
             }
         }
 
-        LOGGER.warn("OIDC delegation failed: {}",
-              failures.stream().collect(Collectors.joining(", ")));
+        LOGGER.warn("OIDC delegation failed: {}", String.join(", ", failures));
         throw new ErrorResponseException(Status.SC_INTERNAL_SERVER_ERROR,
               "Error performing OpenId-Connect delegation");
 
     }
 
-    private HttpPost buildRequest(String token, String host, String clientId, String clientSecret)
-          throws UnsupportedEncodingException, AuthenticationException {
-        UsernamePasswordCredentials clientCreds = new UsernamePasswordCredentials(clientId,
-              clientSecret);
-        BasicScheme scheme = new BasicScheme(UTF_8);
-
-        HttpPost post = new HttpPost(tokenEndPoint(host));
-        List<NameValuePair> params = new ArrayList<>();
-        params.add(new BasicNameValuePair("grant_type", GRANT_TYPE));
-        params.add(new BasicNameValuePair("audience", clientId));
-        params.add(new BasicNameValuePair("subject_token", token));
-        params.add(new BasicNameValuePair("subject_token_type", TOKEN_TYPE));
-        params.add(new BasicNameValuePair("scope", SCOPE));
-
-        post.setEntity(new UrlEncodedFormEntity(params));
-        post.addHeader(scheme.authenticate(clientCreds, post, new BasicHttpContext()));
-        return post;
+    private HttpRequest buildRequest(String token, String host, String clientId, String clientSecret)
+            throws UnsupportedEncodingException{
+
+        return HttpRequest.newBuilder()
+                .uri(URI.create(tokenEndPoint(host)))
+                .POST(
+                        HttpRequest.BodyPublishers.ofString(
+                                String.format(
+                                        "grant_type=%s" +
+                                                "&audience=%s" +
+                                                "&subject_token=%s" +
+                                                "&subject_token_type=%s" +
+                                                "&scope=%s",
+                                        GRANT_TYPE,
+                                        clientId,
+                                        token,
+                                        TOKEN_TYPE,
+                                        SCOPE)
+                        )
+                )
+                .header("Authorization", "Basic " +
+                        Base64.getEncoder().encodeToString(
```

Comment:
This looks too complicated to me. Why not simple:  

```
Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes(UTF_*));
```

Reply:
I didn't know the interaction of "String" + null = "Stringnull", you learn something new every day.

Charset.encode() and String.getBytes() do different things (but should achieve the same result) and I was keeping parity with what Apache HTTP had done. I'll change to getBytes() since from JDK 18 onwards UTF_8 is standard ([See JEP 400](https://openjdk.org/jeps/400)), which means getBytes() is probably quicker thereafter and most likely the same until then.